### PR TITLE
Add Tensorflow and PyTorch Libraries

### LINKED ISSUE #179: Add Tensorflow and PyTorch Libraries

### LINKED ISSUE #213: Remove Auto Types
Overall looks good, I did notice that we are using the `auto` type in a lot of places throughout the codebase and have noted all the locations throughout this PR but will fix the remaining locations throughout the code through a different PR.

_Originally posted by @Byrdman32 in https://github.com/MissouriMRDT/Autonomy_Software/pull/180#pullrequestreview-1953318125_

### PR DESCRIPTION
**Pull request by ClayJay3**
# Main Changes
- Tensorflow and PyTorch Libraries have been added to docker images.

## Other Changes
- Wrote utilities for interfacing with EdgeTPU hardware and loading YOLO models.
- Wrote basic layout for tag and object YOLO detectors.

Resolves #179 
Resolves #213

### CODE REVIEW

File: .devcontainer/Dockerfile_Jammy [hunk 1]
```diff
@@ -28,18 +30,28 @@ RUN echo "${TZ}" > /etc/localtime && \
...
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add - && \
+    curl -fsSL https://bazel.build/bazel-release.pub.gpg | gpg --dearmor >bazel-archive-keyring.gpg && \
+    mv bazel-archive-keyring.gpg /usr/share/keyrings && \
+    echo "deb [arch=amd64 signed-by=/usr/share/keyrings/bazel-archive-keyring.gpg] https://storage.googleapis.com/bazel-apt stable jdk1.8" | tee /etc/apt/sources.list.d/bazel.list
+
 # Install Required Ubuntu Packages
 RUN apt-get update && apt-get install --no-install-recommends -y iputils-ping \
     build-essential gdb wget less udev zstd sudo libgomp1 libswscale-dev \
     cmake git libgtk2.0-dev pkg-config libavcodec-dev libavformat-dev tzdata \
-    libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev tzdata curl net-tools \
```
Comment (Byrdman32) [inline .devcontainer/Dockerfile_Jammy:35]: Why are we getting rid of curl? It's used in the package installer scripts.

File: .devcontainer/Dockerfile_JetPack [hunk 2]
```diff
@@ -22,16 +22,22 @@ ENV LOGNAME root
...
+# Add APT Repo for PCIe drivers and Bazel.
+RUN apt update && apt install -y curl && \
+    echo "deb https://packages.cloud.google.com/apt coral-edgetpu-stable main" | tee /etc/apt/sources.list.d/coral-edgetpu.list && \
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+
 # Install Required Ubuntu Packages
 RUN apt-get update && apt-get install --no-install-recommends -y \
     build-essential gfortran cmake git gdb file tar libatlas-base-dev apt-transport-https iputils-ping \
     libavcodec-dev libavformat-dev libavresample-dev libcanberra-gtk3-module zstd wget less \
-    libdc1394-22-dev libeigen3-dev libglew-dev libgstreamer-plugins-base1.0-dev udev curl net-tools \
```
Comment (Byrdman32) [inline .devcontainer/Dockerfile_JetPack:29]: Why are we getting rid of curl? It's used in our install scripts.

File: examples/vision/cameras/OpenZEDCam.hpp [hunk 3]
```diff
@@ -106,6 +106,7 @@ void RunExample()
         std::future<bool> fuPoseCopyStatus = ExampleZEDCam1->RequestPositionalPoseCopy(slPose);
 
         // Wait for the frames to be copied.
+        // std::this_thread::sleep_for(std::chrono::seconds(1));
```
Comment (Byrdman32) [inline examples/vision/cameras/OpenZEDCam.hpp]: Why do we have this commented out sleep command?

File: src/vision/cameras/ZEDCam.cpp [hunk 4]
```diff
@@ -250,6 +252,9 @@ void ZEDCam::ThreadedContinuousCode()
     }
     else
     {
+        // Record the start time of this block of code.
+        auto tmStartTime = std::chrono::high_resolution_clock::now().time_since_epoch();
```
Comment (Byrdman32) [inline src/vision/cameras/ZEDCam.cpp]: Don't use auto type.

File: tools/action-runners/docker-compose.yml [hunk 5]
```diff
@@ -8,10 +8,10 @@ services:
       context: .
       dockerfile: Dockerfile_Jammy
       args:
-        RUNNER_VERSION: '2.310.2'
+        RUNNER_VERSION: 'latest'
```
Comment (Byrdman32) [inline tools/action-runners/docker-compose.yml]: `latest` doesn't work when I try to create runners. If you want to use latest some more work might be required.

File: tools/action-runners/docker-compose.yml [hunk 6]
```diff
@@ -23,16 +23,17 @@ services:
...
       GH_REPOSITORY: Autonomy_Software
+
   jetpack-runner:
     image: ghcr.io/missourimrdt/jetpack-runner:latest
     build:
       context: .
       dockerfile: Dockerfile_JetPack
       args:
-        RUNNER_VERSION: '2.310.2'
+        RUNNER_VERSION: 'latest'
```
Comment (Byrdman32) [inline tools/action-runners/docker-compose.yml]: `latest` doesn't work when I try to create runners. If you want to use latest some more work might be required.

Comment (Byrdman32) [reply tools/action-runners/docker-compose.yml]: Current is `2.314.1`

File: tools/action-runners/docker-compose.yml (hunk 5 above)
Comment (Byrdman32) [reply tools/action-runners/docker-compose.yml]: Current is `2.314.1`

File: tools/action-runners/scripts/start.sh
Comment (Byrdman32) [inline tools/action-runners/scripts/start.sh:1]: I don't think we are using this anymore but not sure.

File: src/vision/objects/ObjectDetector.cpp [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+            LOG_WARNING(logging::g_qSharedLogger, "ObjectDetector unable to get requested frames from BasicCam!");
+        }
+    }
+
+    /////////////////////////////////////////
+    // Call detection methods and inference.
+    /////////////////////////////////////////
+
+    // Merge the newly detected objects with the pre-existing detected objects
+    // this->UpdateDetectedObjects(vNewlyDetectedObjects);
```
Comment (Byrdman32) [inline src/vision/objects/ObjectDetector.cpp:126]: Is this needed?

**Review by Byrdman32 — CHANGES_REQUESTED** [review]
Overall looks good, I did notice that we are using the `auto` type in a lot of places throughout the codebase and have noted all the locations throughout this PR but will fix the remaining locations throughout the code through a different PR.

File: .devcontainer/Dockerfile_Jammy (hunk 1 above)
Comment (ClayJay3) [reply .devcontainer/Dockerfile_Jammy:35]: Curl is already installed in the base image. It also gets reinstalled in the Dockerfiles for the package runners. So I removed it as cleanup.

File: src/vision/cameras/ZEDCam.cpp (hunk 4 above)
Comment (ClayJay3) [reply src/vision/cameras/ZEDCam.cpp]: I went through the project and removed all instances of auto and replaced them with proper types where possible.

File: src/vision/objects/ObjectDetector.cpp (hunk 7 above)
Comment (ClayJay3) [reply src/vision/objects/ObjectDetector.cpp:126]: Added TODO for clarity, This will be added in the future.

File: tools/action-runners/scripts/start.sh
Comment (ClayJay3) [reply tools/action-runners/scripts/start.sh:1]: This doesn't exist on my end, not sure why this is still here.

File: examples/vision/cameras/OpenZEDCam.hpp (hunk 3 above)
Comment (ClayJay3) [reply examples/vision/cameras/OpenZEDCam.hpp]: Removed old line used for debugging.

**Review by Byrdman32 — APPROVED** [review]

File: .devcontainer/devcontainer.json [hunk 8]
```diff
@@ -30,11 +30,11 @@
     "mounts": [
         "source=${localWorkspaceFolder}/data/models/zed,target=/usr/local/zed/resources,type=bind,consistency=delegated"
     ],
-    "image": "ghcr.io/missourimrdt/autonomy-jammy:latest",
```
Comment (Byrdman32) [inline .devcontainer/devcontainer.json:33]: And this needs to be set to latest once we have a working image

Comment (ClayJay3) [reply .devcontainer/devcontainer.json:33]: I went ahead and set it to latest, all latest images should be working once the actions finish.

**Pull request merged**